ibis/common/patterns.py

```diff
@@ -1753,6 +1782,39 @@ def FrozenDictOf(key_pattern, value_pattern):
     return MappingOf(key_pattern, value_pattern, type=frozendict)
 
 
+def builder(obj):
+    """Convert an object to a builder.
+
+    It encapsulates:
+        - callable objects into a `Factory` builder
+        - non-callable objects into a `Just` builder
+
+    Parameters
+    ----------
+    obj
+        The object to convert to a builder.
+
+    Returns
+    -------
+    The builder instance.
+    """
+    if isinstance(obj, Builder):
+        # already a builder, no need to convert
+        return obj
+    elif callable(obj):
+        # the callable builds the substitution
+        return Factory(obj)
+    elif isinstance(obj, Mapping):
+        # allow nesting builder patterns in dicts
+        return Call(type(obj), **obj)
+    elif is_iterable(obj):
```

Comment:
This is handling the empty tuple/list/etc case?

Reply:
Yes, added tests to cover them.